cpp/test/reader/tsfile_reader_test.cc

```diff
@@ -0,0 +1,221 @@
+            "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+        std::string random_string;
+
+        for (int i = 0; i < length; ++i) {
+            random_string += chars[dis(gen)];
+        }
+
+        return random_string;
+    }
+
+    static std::string field_to_string(storage::Field *value) {
+        if (value->type_ == common::TEXT) {
+            return std::string(value->value_.sval_);
+        } else {
+            std::stringstream ss;
+            switch (value->type_) {
+                case common::BOOLEAN:
+                    ss << (value->value_.bval_ ? "true" : "false");
+                    break;
+                case common::INT32:
+                    ss << value->value_.ival_;
+                    break;
+                case common::INT64:
+                    ss << value->value_.lval_;
+                    break;
+                case common::FLOAT:
+                    ss << value->value_.fval_;
+                    break;
+                case common::DOUBLE:
+                    ss << value->value_.dval_;
+                    break;
+                case common::NULL_TYPE:
+                    ss << "NULL";
+                    break;
+                default:
+                    ASSERT(false);
+                    break;
+            }
+            return ss.str();
+        }
+    }
+};
+
+TEST_F(TsFileReaderTest, ResultSetMetadata) {
+    std::string device_path = "device1";
+    std::string measurement_name = "temperature";
+    common::TSDataType data_type = common::TSDataType::INT32;
+    common::TSEncoding encoding = common::TSEncoding::PLAIN;
+    common::CompressionType compression_type =
+        common::CompressionType::UNCOMPRESSED;
+    tsfile_writer_->register_timeseries(
+        device_path, storage::MeasurementSchema(measurement_name, data_type,
+                                                encoding, compression_type));
+
+    for (int i = 0; i < 50000; ++i) {
+        TsRecord record(1622505600000 + i * 1000, device_path);
+        record.add_point(measurement_name, (int32_t)i);
+        ASSERT_EQ(tsfile_writer_->write_record(record), E_OK);
+        ASSERT_EQ(tsfile_writer_->flush(), E_OK);
+    }
+    ASSERT_EQ(tsfile_writer_->close(), E_OK);
+
+    std::vector<std::string> select_list = {"device1.temperature"};
+
+    storage::TsFileReader reader;
+    int ret = reader.open(file_name_);
+    ASSERT_EQ(ret, common::E_OK);
+    storage::ResultSet *tmp_qds = nullptr;
+
+    ret = reader.query(select_list, 1622505600000, 1622505600000 + 50000 * 1000,
+                       tmp_qds);
+    auto *qds = (QDSWithoutTimeGenerator *)tmp_qds;
+
+    ResultSetMetadata *result_set_metadaa = qds->get_metadata();
+    ASSERT_EQ(result_set_metadaa->get_column_type(0), data_type);
+    ASSERT_EQ(result_set_metadaa->get_column_name(0),
+              device_path + "." + measurement_name);
+    reader.destroy_query_data_set(qds);
+    reader.close();
+}
+
+TEST_F(TsFileReaderTest, GetAllDevice) {
```

Comment:
Add tests with enough devices/measurements to generate internal nodes in the index.